jerry-core/ecma/operations/ecma-objects-general.c

```diff
@@ -265,179 +292,147 @@ ecma_op_general_object_put (ecma_object_t *obj_p, /**< the object */
+        {
+          setter_p = ecma_get_named_accessor_property_setter (inherited_prop_p);
+          create_new_property = false;
+        }
+        else
+        {
+          create_new_property = ecma_is_property_writable (inherited_prop_p);
+        }
+      }
+    }
 
-  // 2.
-  if (prop_p != NULL)
-  {
-    // a.
-    if (ECMA_PROPERTY_GET_TYPE (prop_p) == ECMA_PROPERTY_TYPE_NAMEDACCESSOR)
+    if (create_new_property
+        && ecma_get_object_extensible (obj_p))
     {
-      ecma_object_t *setter_p = ecma_get_named_accessor_property_setter (prop_p);
+      const ecma_object_type_t type = ecma_get_object_type (obj_p);
 
-      // i.
-      if (setter_p == NULL)
+      if (type == ECMA_OBJECT_TYPE_ARGUMENTS)
       {
-        return false;
+        return ecma_builtin_helper_def_prop (obj_p,
+                                             property_name_p,
+                                             value,
+                                             true, /* Writable */
+                                             true, /* Enumerable */
+                                             true, /* Configurable */
+                                             is_throw); /* Failure handling */
       }
 
-      // ii.
-      return true;
-    }
-    else
-    {
-      // b.
-      JERRY_ASSERT (ECMA_PROPERTY_GET_TYPE (prop_p) == ECMA_PROPERTY_TYPE_NAMEDDATA);
+      uint32_t index;
 
-      return ecma_is_property_writable (prop_p);
-    }
-  }
+      if (type == ECMA_OBJECT_TYPE_ARRAY
+          && ecma_string_get_array_index (property_name_p, &index))
+      {
+        /* Since the length of an array is a non-configurable named data
+         * property, the prop_p must be a non-NULL pointer for all arrays. */
 
-  // 3.
-  ecma_object_t *proto_p = ecma_get_object_prototype (obj_p);
+        JERRY_ASSERT (!ecma_op_general_object_property_name_is_length (property_name_p));
 
-  // 4.
-  if (proto_p == NULL)
-  {
-    return ecma_get_object_extensible (obj_p);
-  }
+        ecma_string_t *magic_string_length_p = ecma_get_magic_string (LIT_MAGIC_STRING_LENGTH);
+        ecma_property_t *len_prop_p = ecma_op_object_get_own_property (obj_p, magic_string_length_p);
+        ecma_deref_ecma_string (magic_string_length_p);
 
-  // 5.
-  ecma_property_t *inherited_p = ecma_op_object_get_property (proto_p, property_name_p);
+        JERRY_ASSERT (len_prop_p != NULL
+                      && ECMA_PROPERTY_GET_TYPE (len_prop_p) == ECMA_PROPERTY_TYPE_NAMEDDATA);
 
-  // 6.
-  if (inherited_p == NULL)
-  {
-    return ecma_get_object_extensible (obj_p);
-  }
+        uint32_t old_len = ecma_get_uint32_from_value (ecma_get_named_data_property_value (len_prop_p));
 
-  // 7.
-  if (ECMA_PROPERTY_GET_TYPE (inherited_p) == ECMA_PROPERTY_TYPE_NAMEDACCESSOR)
-  {
-    ecma_object_t *setter_p = ecma_get_named_accessor_property_setter (inherited_p);
+        if (index < UINT32_MAX
```

Comment:
What happens if `index == UINT32_MAX`? Is it possible to reach the maximum value?